Compatibility Fix: Do not load translations early

RadoslavGeorgiev opened the pull request:
Fixes #9727 

#### Changes proposed in this Pull Request

Based on https://github.com/Automattic/woocommerce-payments/pull/9884, to avoid also seeing Jetpack initialization warnings. 

This PR is an alternative approach to @mordeth's idea to shuffle initialization, https://github.com/Automattic/woocommerce-payments/pull/9872. However, instead of re-ordering the whole initialization project, which seems a bit risky, this PR focuses on only loading translations at the right time.

#### Testing instructions

1. Switch to a non-English version for WordPress. This will trigger the loading of translations.
2. Check out this branch.
3. Navigate the front- and back-end, perform various actions (ex. checkout).
4. Ensure that there are no warnings regarding `woocommerce-payments` translations being loaded too early. 

⚠️ You are still going to see "Function `_load_textdomain_just_in_time` was called incorrectly." warnings, but they should not be related to the `woocommerce-payments` domain.

-------------------

- [x] Run `npm run changelog` to add a changelog file, choose `patch` to leave it empty if the change is not significant. You can add multiple changelog files in one PR by running this command a few times. 
- [x] Covered with tests (or have a good reason not to test in description ☝️)

**Post merge**

<!--
Make sure you edit the page for the current release when adding testing instructions.
We often create a blank page ahead of time for the next release.
If this PR need not be QA tested, edit to 'QA Testing Not Applicable'
-->

- [ ] Link to testing instructions from [release testing doc](https://github.com/Automattic/woocommerce-payments/wiki/Release-testing-instructions) following [these instructions](https://github.com/Automattic/woocommerce-payments/wiki/How-to-write-good-manual-testing-scenarios) : _Add link here / 'QA Testing Not Applicable'_
- [ ] Add or update [critical flows](https://github.com/Automattic/woocommerce-payments/wiki/Critical-flows) and [testing instructions for critical flows](https://github.com/Automattic/woocommerce-payments/wiki/Testing-instructions-for-critical-flows), if applicable.
- [ ] Add what's changed (description, screenshot, demo videos etc.) to the release announcement post, if applicable.

zinigor committed:
Added a callback instead of a direct call to avoid notices.

zinigor committed:
Changelog.

zinigor committed:
Update woocommerce-payments.php

Co-authored-by: Cvetan Cvetanov <cvetan.cvetanov@automattic.com>

zinigor committed:
Updated package versions.

RadoslavGeorgiev committed:
Initial changes

botwoo commented:
### Test the build 

#### Option 1. Jetpack Beta

- Install and activate [Jetpack Beta](https://jetpack.com/download-jetpack-beta/).
- Use this build by searching for PR number `9940` or branch name `compat/9727-avoid-early-translations` in your-test.site/wp-admin/admin.php?page=jetpack-beta&plugin=woocommerce-payments          

#### Option 2. Jurassic Ninja - available for logged-in A12s

:rocket: [Launch a JN site with this branch](https://jurassic.ninja/create/?jetpack-beta&shortlived&nojetpack&woocommerce&woocommerce-payments-dev-tools&branches.woocommerce-payments=compat/9727-avoid-early-translations) :rocket:

:information_source: Install this [Tampermonkey script](https://github.com/Automattic/woocommerce-payments/tree/develop/bin/wcpay-live-branches) to get more options.

---

Build info: 

- Latest commit: b2e41e09bd7fd14e892c2e48fab3b5e57d866d84
- Build time: 2024-12-12 22:12:32 UTC

_Note: the build is updated when a new commit is pushed to this PR._

RadoslavGeorgiev committed:
Restoring subs core

RadoslavGeorgiev committed:
Move titles of payment methods to a separate method

RadoslavGeorgiev committed:
Extract gateway fields into a separate method

RadoslavGeorgiev committed:
Fix how the method description is set

The core payment gateway class only uses the `method_description` property within `get_method_description()`, a method that we are already overloading. This makes the `$this->method_description = $this->get_method_description()` call redundant, especially in the constructor.

RadoslavGeorgiev committed:
Changelog entry

RadoslavGeorgiev committed:
Fix tests

The payment admin's `add_payments_menu` method now needs to be called in tests, to simulate the real environment.

RadoslavGeorgiev committed:
Extract the title as a method in the main gateway class

zmaglica commented on includes/admin/class-wc-payments-admin.php:
```diff
@@ -315,6 +272,49 @@ public function add_payments_menu() {
 		}
 		global $submenu;
 
+		$this->admin_child_pages = [
```
Nitpick. Feel free to ignore. What if we move the logic for `$this->admin_child_pages` into a separate function like `get_admin_child_pages()` and call it whenever needed? Currently, `$this->admin_child_pages` is populated in `add_payments_menu` and used in both `add_payments_menu` and `maybe_redirect_from_payments_admin_child_pages`. There’s a possibility (or maybe not) that `$this->admin_child_pages` could be empty in the redirect function if it hasn’t been populated yet, causing redirects to fail.

That said, since `maybe_redirect_from_payments_admin_child_pages` is hooked to the `admin_init` action, there’s a chance this change might reintroduce the issue that this PR was intended to solve. What are your thoughts on moving `$this->admin_child_pages` logic into `get_admin_child_pages()` and calling it directly from the redirect function?

RadoslavGeorgiev replied:
I considered this, but decided to keep it as-is.

> Currently, $this->admin_child_pages is populated in add_payments_menu and used in both add_payments_menu and maybe_redirect_from_payments_admin_child_pages

I'd just like to mention that this is not something that is getting introduced within this PR, but has been this way for a while, since the final `admin_child_pages` is only available after `add_payments_menu` runs. That's because the method adds and modifies pages.

> There’s a possibility (or maybe not) that $this->admin_child_pages could be empty in the redirect function if it hasn’t been populated yet, causing redirects to fail.

The idea of the method is to redirect away from a page, if the user does not access it. There is technically no page before the `admin_menu` 

> That said, since `maybe_redirect_from_payments_admin_child_pages` is hooked to the `admin_init` action, there’s a chance this change might reintroduce the issue that this PR was intended to solve. 

That's another +1 for keeping it as is: It requires `maybe_redirect_from_payments_admin_child_pages` to run late, and prevents it from meaningfully running before translations are properly loaded.

That's to say, I'm merging the PR with the current structure.

zmaglica reviewed (approved):
Left one comment. It is a nitpicky one. Feel free to ignore it. Overall, good work on this PR.

mordeth reviewed (approved):
The code changes look excellent.

I fully agree that this is a much-better approach to handling the translations fix.

mgascam reviewed (approved):
Code changes look good to me and I love the approach. I tested setting the language to Spanish and everything works well. Also, no more instances of `Translation loading for the <code>woocommerce-payments</code> domain was triggered too early.` in the logs with this branch. 🚢
